File: bin/day-11.py
import re
import string
import time


valid_triplets = ["%s%s%s" % (c, chr(ord(c)+1), chr(ord(c)+2)) for c in string.lowercase[:-2]]
pair_re = re.compile(r'(\w)\1.*(\w)\2')


def increment(pw):
    last = pw[-1]
    if last == 'z':
        return increment(pw[:-1]) + 'a'
    return pw[:-1] + chr(ord(last)+1)


def is_valid(pw):
    if 'i' in pw or 'o' in pw or 'l' in pw:
        return False

    if not pair_re.search(pw):
        return False

    for vt in valid_triplets:
        if vt in pw:
            break
    else:
        return False
    return True


password = "hepxcrrq"


now = time.time()
while not is_valid(password):
    password = increment(password)
print password, " took ", int((time.time() - now) * 1000), 'ms'

now = time.time()
password = increment(password)
while not is_valid(password):
    password = increment(password)
print password, " took ", int((time.time() - now) * 1000), 'ms'

File: bin/day-11.dart
// This is a direct port of the python version.
// the main difference is the use of String.fromCharCode and String.codeUnits
// which seems like a necessary evil of string manipulation in dart.
RegExp pairsPattern = new RegExp(r'(\w+)\1.*(\w)\2');

List<String> triplets = 'abcdefghijklmnopqrstuvwx'.split('').map(
    (String c) => (c + new String.fromCharCode(c.codeUnits[0]+1)
                   + new String.fromCharCode(c.codeUnits[0]+2))).toList();

String increment(String input) {
  int last = input.codeUnits.last;
  if (last == 122) {
    return increment(input.substring(0, input.length-1)) + 'a';
  } else {
    return input.substring(0, input.length-1) + new String.fromCharCode(last+1);
  }
}

bool isValid(String pw) {
  if (pw.contains('l') || pw.contains('i') || pw.contains('o')) {
    return false;
  }
  if (!pairsPattern.hasMatch(pw)) {
    return false;
  }
  bool found = false;
  for (String trip in triplets) {
    if (pw.contains(trip)) {
      found = true;
      break;
    }
  }
  return found;
}

main() {
  Stopwatch watch = new Stopwatch()..start();
  String password = 'hepxcrrq';
  while (!isValid(password)) {
    password = increment(password);
  }
  watch.stop();
  print("Next Password is $password (took ${watch.elapsedMilliseconds} ms)");
  watch..reset()..start();
  password = increment(password);
  while(!isValid(password)) {
    password = increment(password);
  }
  watch.stop();
  print("Next Password is $password (took ${watch.elapsedMilliseconds} ms)");
}